### PR TITLE
Add stable_bl_edmf_implicit.jl example, which uses fully implicit solver to solve EDMF

### PR DESCRIPTION
**Pull request by Zhengyu-Huang**
This is an example which uses fully implicit solver to solve EDMF(stable boundary layer) for 6h within 15 mins.
This example only demonstrates that we can use fully implicit solver here, but I have not visualized/compared the results

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by ilopezgp** [comment]
This test should be added to the `.buildkite/pipeline.yml`.

**Review by ilopezgp — APPROVED** [review]
LGTM, I think the test has to be buildkite.

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        end
+
+    cb_check_cons = GenericCallbacks.EveryXSimulationSteps(3000) do
+        Q = solver_config.Q
+        δρ = (sum(Q.ρ .* M) - Σρ₀) / Σρ₀
+        δρe = (sum(Q.ρe .* M) .- Σρe₀) ./ Σρe₀
+        @show (abs(δρ))
+        @show (abs(δρe))
+        @warn (abs(δρ) <= 0.001)
+        @warn (abs(δρe) <= 0.0025)
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/stable_bl_edmf_implicit.jl:288]: Shouldn't these be `@show` s?

File: src/Numerics/SystemSolvers/SystemSolvers.jl [hunk 2]
```diff
@@ -151,7 +151,7 @@ function nonlinearsolve!(
         # ||Delta Q|| / ||Q|| ?
         relresidual = residual_norm / initial_residual_norm
         if relresidual < tol || residual_norm < tol
-            @info "Newton converged in $iters iterations!"
+            # @info "Newton converged in $iters iterations!"
```
Comment (charleskawczynski) [inline src/Numerics/SystemSolvers/SystemSolvers.jl:154]: Should we just delete these `@info`s?

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+            FT(1 - z / 250.0) *
+            FT(1 - z / 250.0) *
+            FT(1 - z / 250.0)
+        en.ρaθ_liq_cv =
+            gm.ρ *
+            FT(0.4) *
+            FT(1 - z / 250.0) *
+            FT(1 - z / 250.0) *
+            FT(1 - z / 250.0)
+    else
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/stable_bl_edmf_implicit.jl:79]:
```suggestion
    if z <= FT(250)
        coeff = gm.ρ *
            FT(0.4) *
            FT(1 - z / 250.0) *
            FT(1 - z / 250.0) *
            FT(1 - z / 250.0)
        en.ρatke = coeff
        en.ρaθ_liq_cv = coeff
    else
```

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl (hunk 1 above)
Comment (Zhengyu-Huang) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl:288]: I am not sure about this, originally these are error(), I do not know the purpose @ilopezgp

File: src/Numerics/SystemSolvers/SystemSolvers.jl (hunk 2 above)
Comment (Zhengyu-Huang) [reply src/Numerics/SystemSolvers/SystemSolvers.jl:154]: Still plan to leave these commented out info here, and we need to use them to help tune parameters for new cases

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    Q = solver_config.Q
+    # Volume geometry information
+    vgeo = driver_config.grid.vgeo
+    M = vgeo[:, Grids._M, :]
+    # Unpack prognostic vars
+    ρ₀ = Q.ρ
+    ρe₀ = Q.ρe
+    # DG variable sums
+    Σρ₀ = sum(ρ₀ .* M)
+    Σρe₀ = sum(ρe₀ .* M)
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/stable_bl_edmf_implicit.jl:252]: Can we use the new way of adding conservation test callbacks? (you can check other experiments/drivers, they do not use `vgeo`)

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl (hunk 1 above)
Comment (charleskawczynski) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl:283]: No, originally they were `@test`, which ensures that it passes the test. When running this earlier, I saw that this fails the conservation tests.

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl (hunk 4 above)
Comment (jkozdon) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl:252]: [`weightedsum`](https://github.com/CliMA/ClimateMachine.jl/blob/0e9f0b985eedc1e2168a2a24a126265555424a6b/src/Arrays/MPIStateArrays.jl#L630-L638) should probably be used here.

This is an example of usage: https://github.com/CliMA/ClimateMachine.jl/blob/7d034daf7bae3e9a7131546517dca54903d26806/src/Diagnostics/atmos_mass_energy_loss.jl#L65-L68

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl (hunk 1 above)
Comment (Zhengyu-Huang) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl:283]: Oh got it, yes, the energy has has relative error larger than 0.0025

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        arg_type = String
+        default = "bulk"
+    end
+
+    cl_args = ClimateMachine.init(parse_clargs = true, custom_clargs = sbl_args)
+
+    surface_flux = cl_args["surface_flux"]
+
+    # DG polynomial order
+    N = 1
```
Comment (charleskawczynski) [inline test/Atmos/EDMF/stable_bl_edmf_implicit.jl:101]: We should really set `N = 3` at least. I know that it's `N = 1` in master, but we shouldn't have that, and we should fix it here.

Comment (charleskawczynski) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl:101]: We can reduce `nelem_vert` to offset this a bit.

Comment (Zhengyu-Huang) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl:104]: will revisit the vertical spatial discretization after the horizontal-vertical splitting is done, I think a stabler spatial discretization will help the convergence

File: src/Atmos/Model/AtmosModel.jl [hunk 6]
```diff
@@ -237,16 +237,23 @@ end
     vars_state(m::AtmosModel, ::Prognostic, FT)
 
 Conserved state variables (prognostic variables).
+
+!!! warning
+
+    The order of the fields for `AtmosModel` needs to match the one for
+    `AtmosLinearModel` since a shared state is used
```
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:244]: There is a rebase conflict

**Review by charleskawczynski — CHANGES_REQUESTED** [review]
Need to rebase

File: src/Atmos/Model/AtmosModel.jl (hunk 6 above)
Comment (Zhengyu-Huang) [reply src/Atmos/Model/AtmosModel.jl:244]: just rebased

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:244]: This code already exists in master, so I think something is still wrong.

File: test/Atmos/EDMF/stable_bl_edmf_implicit.jl (hunk 3 above)
Comment (Zhengyu-Huang) [reply test/Atmos/EDMF/stable_bl_edmf_implicit.jl]: Does this commit  introduce "Fix the ordering of the state in the AtmosModel" and "pass param and time to preconditioner_update" ?

File: src/Atmos/Model/AtmosModel.jl (hunk 6 above)
Comment (Zhengyu-Huang) [reply src/Atmos/Model/AtmosModel.jl:244]: I will close this one and do a new PR

**Pull request closed without merging**